File: find_closest_value_in_bst/find_closest_value_in_bst.ts
class BST {
  value: number;
  left: BST | null = null;
  right: BST | null = null;

  constructor(value: number) {
      this.value = value;
  }
}

function findClosestValueInBST(tree: BST, target: number): number {
  let currentNode: BST | null = tree;
  let closestValue = currentNode.value;

  while (currentNode !== null) {
      if (Math.abs(currentNode.value - target) < Math.abs(closestValue - target)) {
          closestValue = currentNode.value;
      }
      if (target > currentNode.value) {
          currentNode = currentNode.right;
      } else {
          currentNode = currentNode.left;
      }
  }

  return closestValue;
}

const bst = new BST(10);
bst.left = new BST(8);
bst.right = new BST(12);

console.log(findClosestValueInBST(bst, 20));

File: find_closest_value_in_bst/find_closest_value_in_bst.java
public class BST {
    int value;
    BST left;
    BST right;

    BST(int value) {
        this.value = value;
    }

    public static int findClosestValueInBST(BST tree, int target) {
        BST currentNode = tree;
        int closestValue = currentNode.value;

        while (currentNode != null) {
            if (Math.abs(currentNode.value - target) < Math.abs(closestValue - target)) {
                closestValue = currentNode.value;
            }
            if (target > currentNode.value) {
                currentNode = currentNode.right;
            } else {
                currentNode = currentNode.left;
            }
        }
        return closestValue;
    }

    public static void main(String[] args) {
        BST bst = new BST(10);
        bst.left = new BST(8);
        bst.right = new BST(12);

        int result = findClosestValueInBST(bst, 20);
        System.out.println(result);  
    }
}
